dot/sync/syncer.go

```diff
@@ -424,6 +439,27 @@ func (s *Service) handleRuntimeChanges(newState *rtstorage.TrieState) error {
 		return ErrEmptyRuntimeCode
 	}
 
+	codeSubBlockHash, err := s.codeSubstitutedState.LoadCodeSubstitutedBlockHash()
+	if err != nil {
+		logger.Debug("problem checking code substituted block hash", "error", err)
+	}
+
+	if !reflect.DeepEqual(codeSubBlockHash, common.Hash{}) {
+		// don't do runtime change if using code substitution and runtime change spec version are equal
+		//  (do a runtime change if code substituted and runtime spec versions are different, or code not substituted)
+		newVersion, err := s.runtime.CheckRuntimeVersion(code) // nolint
+		if err != nil {
+			logger.Debug("problem checking runtime version", "error", err)
+		}
+
+		if previousVersion.SpecVersion() == newVersion.SpecVersion() {
+			return nil
+		}
+		logger.Info("🔄 detected runtime code change, upgrading...", "block", s.blockState.BestBlockHash(),
+			"previous code hash", s.codeHash, "new code hash", currCodeHash,
+			"previous spec version", previousVersion.SpecVersion(), "new spec version", newVersion.SpecVersion())
+	}
+
```

Comment:
you will also need to do `StoreCodeSubstitutedBlockHash(common.Hash{})` here to store in the db that no code substitution is currently happening, otherwise on restart the node will still think there's a code substitution

Reply:
Good catch, updated.